src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts

```diff
@@ -0,0 +1,386 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { setup_proxy_console } from "../logging";
+import { InitCryptoMessageData } from "../crypto-worker";
+
+class FailedOrStoppedLoopError extends Error { }
+class ArgumentsError extends Error { }
+class WorkerFailedError extends Error { }
+
+class ChannelWorker {
+    // LOCK states
+    get LOCK_UNLOCKED() { return 0; } // 0 means the lock is unlocked
+    get LOCK_OWNED() { return 2; } // 2 means the ChannelWorker owns the lock
+
+    // BEGIN ChannelOwner contract - shared constants.
+    get STATE_IDX() { return 0; }
+    get MSG_SIZE_IDX() { return 1; }
+    get LOCK_IDX() { return 2; }
+
+    // Communication states.
+    get STATE_SHUTDOWN() { return -1; } // Shutdown
+    get STATE_IDLE() { return 0; }
+    get STATE_REQ() { return 1; }
+    get STATE_RESP() { return 2; }
+    get STATE_REQ_P() { return 3; } // Request has multiple parts
+    get STATE_RESP_P() { return 4; } // Response has multiple parts
+    get STATE_AWAIT() { return 5; } // Awaiting the next part
+    get STATE_REQ_FAILED() { return 6; } // The Request failed
+    get STATE_RESET() { return 7; } // Reset to a known state
+    // END ChannelOwner contract - shared constants.
+
+    private comm: Int32Array;
+    private msg: Uint16Array;
+    private msg_char_len: number;
+
+    constructor(comm_buf: SharedArrayBuffer, msg_buf: SharedArrayBuffer, msg_char_len: number) {
+        this.comm = new Int32Array(comm_buf);
+        this.msg = new Uint16Array(msg_buf);
+        this.msg_char_len = msg_char_len;
+    }
+
+    async run_message_loop(async_op: (jsonRequest: string) => Promise<number[]>) {
+        for (; ;) {
+            try {
+                // Wait for signal to perform operation
+                let state;
+                do {
+                    this._wait(this.STATE_IDLE);
+                    state = Atomics.load(this.comm, this.STATE_IDX);
+                } while (state !== this.STATE_REQ && state !== this.STATE_REQ_P && state !== this.STATE_SHUTDOWN && state !== this.STATE_REQ_FAILED && state !== this.STATE_RESET);
+
+                this._throw_if_reset_or_shutdown();
+
+                // Read in request
+                const req: string = this._read_request();
```

Comment:
Does this need the explicit type? It should be inferred from the method it calls.

```suggestion
                const req = this._read_request();
```